medmodels/matching/metrics.py

```diff
@@ -63,6 +67,49 @@ def exact_metric(vector1: NDArray[np.float64], vector2: NDArray[np.float64]) ->
     return np.inf
 
 
+def mahalanobis_metric(
+    vector1: NDArray[np.float64],
+    vector2: NDArray[np.float64],
+    inv_cov: NDArray[np.float64],
+):
+    """
+    Returns mahalanobis metric for matching. Works better with continuous covariates.
+
+    $$
+    D(x, y) = \\sqrt{(x-y)^T S^{-1} (x-y)} \\\\
+    \\text{where } S \\text{ is the covariance matrix of the whole distribution}
+    $$
+
+    Args:
+        vector1 (NDArray[np.float64]): The first vector to be compared.
+        vector2 (NDArray[np.float64]): The second vector to be compared. Must have the
+            same shape as `vector1` and belong to the same distribution.
+        inv_cov (NDArray[np.float64]): The inverse of the covariance matrix of the
+            whole distribution (data set).
+
+    Returns:
+        float: The Mahalanobis distance between the two vectors.
+
+    The covariance matrix and its inverse are calculated at most ones per item to be
+    paired, hence, they won't be included inside of the method in order to avoid the
+    repeated computation.
+
+    By matching without replacement the found paired item will be removed from the set,
+    hence,the covariance matrix of the whole distribution and its inverse need to be
+    recalculated for every entry. This can be time consuming for big data sets (esp.
+    with a big amount of features).
+    """
+
+    diff = vector1 - vector2
+    left = np.dot(diff, inv_cov)
```

Comment:
I researched the np.dot function a bit, because I was a bit confused by the missing transpose in this. I read that np.dot can have a bit of a confusing behaviour and using matrix multiplication operator @ and explicitly using the transpose could be more clear.

Reply:
Very good comment! Thanks. Now I know more about this!

Reply:
From this comment, I realized that having a more clear mahalanobis function would be best. That is, including the difference between what happens in the three scenarios:
- 1-D (we do not use covariance really, but standard deviation)
- 2+-D (normal mahalanobis)
- Singular covariance matrix: in some cases, the inverse covariance cannot be computed since the matrix is singular, and an error should apper.